File: packages/js/src/extensions/ai-ml.ts
/**
 * AI/ML Extensions for JSON-LD
 *
 * Provides support for:
 * - @confidence: Assertion confidence scores (0.0 - 1.0)
 * - @source: Provenance tracking (model/system IRI)
 * - @extractedAt: Extraction timestamps
 * - @method: Extraction method descriptions
 * - @humanVerified: Human verification flags
 *
 * These extensions enable JSON-LD to serve as a standard interchange
 * format for AI/ML-generated knowledge graphs with full provenance.
 */

import {
  KEYWORD_CONFIDENCE,
  KEYWORD_SOURCE,
  KEYWORD_EXTRACTED_AT,
  KEYWORD_METHOD,
  KEYWORD_HUMAN_VERIFIED,
  KEYWORD_MEDIA_TYPE,
  KEYWORD_CONTENT_URL,
  KEYWORD_CONTENT_HASH,
  KEYWORD_TRANSLATED_FROM,
  KEYWORD_TRANSLATION_MODEL,
  KEYWORD_MEASUREMENT_UNCERTAINTY,
  KEYWORD_UNIT,
  KEYWORD_ACTED_ON_BEHALF_OF,
  KEYWORD_WAS_DERIVED_FROM,
  KEYWORD_WAS_INVALIDATED_BY,
  KEYWORD_CALIBRATION_STATUS,
  KEYWORD_CALIBRATION_TARGET,
  KEYWORD_WAS_AGGREGATED_BY,
  JSONLD_EX_NAMESPACE,
} from '../keywords.js';
import { AnnotatedValue, ProvenanceMetadata } from '../types.js';

// ── Context Definition ────────────────────────────────────────────

/**
 * JSON-LD context fragment that defines the AI/ML extension terms.
 * Merge this into your document's @context to enable provenance annotations.
 */
export const AI_ML_CONTEXT = {
  '@vocab': JSONLD_EX_NAMESPACE,
  'confidence': {
    '@id': `${JSONLD_EX_NAMESPACE}confidence`,
    '@type': 'http://www.w3.org/2001/XMLSchema#double',
  },
  'source': {
    '@id': `${JSONLD_EX_NAMESPACE}source`,
    '@type': '@id',
  },
  'extractedAt': {
    '@id': `${JSONLD_EX_NAMESPACE}extractedAt`,
    '@type': 'http://www.w3.org/2001/XMLSchema#dateTime',
  },
  'method': {
    '@id': `${JSONLD_EX_NAMESPACE}method`,
    '@type': 'http://www.w3.org/2001/XMLSchema#string',
  },
  'humanVerified': {
    '@id': `${JSONLD_EX_NAMESPACE}humanVerified`,
    '@type': 'http://www.w3.org/2001/XMLSchema#boolean',
  },
  'mediaType': {
    '@id': `${JSONLD_EX_NAMESPACE}mediaType`,
    '@type': 'http://www.w3.org/2001/XMLSchema#string',
  },
  'contentUrl': {
    '@id': `${JSONLD_EX_NAMESPACE}contentUrl`,
    '@type': '@id',
  },
  'contentHash': {
    '@id': `${JSONLD_EX_NAMESPACE}contentHash`,
    '@type': 'http://www.w3.org/2001/XMLSchema#string',
  },
  'translatedFrom': {
    '@id': `${JSONLD_EX_NAMESPACE}translatedFrom`,
    '@type': '@id', // or string depending on usage, usually IRI
  },
  'translationModel': {
    '@id': `${JSONLD_EX_NAMESPACE}translationModel`,
    '@type': 'http://www.w3.org/2001/XMLSchema#string',
  },
  'measurementUncertainty': {
    '@id': `${JSONLD_EX_NAMESPACE}measurementUncertainty`,
    '@type': 'http://www.w3.org/2001/XMLSchema#double',
  },
  'unit': {
    '@id': `${JSONLD_EX_NAMESPACE}unit`,
    '@type': 'http://www.w3.org/2001/XMLSchema#string',
  },
  'actedOnBehalfOf': {
    '@id': `${JSONLD_EX_NAMESPACE}actedOnBehalfOf`,
    '@type': '@id',
  },
  'wasDerivedFrom': {
    '@id': `${JSONLD_EX_NAMESPACE}wasDerivedFrom`,
    '@type': '@id',
  },
  'wasInvalidatedBy': {
    '@id': `${JSONLD_EX_NAMESPACE}wasInvalidatedBy`,
    '@type': '@id',
  },
  'calibrationStatus': {
    '@id': `${JSONLD_EX_NAMESPACE}calibrationStatus`,
    '@type': 'http://www.w3.org/2001/XMLSchema#string',
  },
  'calibrationTarget': {
    '@id': `${JSONLD_EX_NAMESPACE}calibrationTarget`,
    '@type': 'http://www.w3.org/2001/XMLSchema#string',
  },
  'wasAggregatedBy': {
    '@id': `${JSONLD_EX_NAMESPACE}wasAggregatedBy`,
    '@type': '@id',
  },
};

// ── Annotation Builder ────────────────────────────────────────────

/**
 * Creates an annotated value with AI/ML provenance metadata.
 *
 * @example
 * ```ts
 * const name = annotate("John Smith", {
 *   confidence: 0.95,
 *   source: "https://ml-model.example.org/ner-v2",
 *   method: "NER",
 * });
 * // Returns: { "@value": "John Smith", "@confidence": 0.95, ... }
 * ```
 */
export function annotate(
  value: string | number | boolean,
  metadata: ProvenanceMetadata
): AnnotatedValue {
  const result: AnnotatedValue = { '@value': value };

  if (metadata.confidence !== undefined) {
    validateConfidence(metadata.confidence);
    result['@confidence'] = metadata.confidence;
  }
  if (metadata.source !== undefined) {
    result['@source'] = metadata.source;
  }
  if (metadata.extractedAt !== undefined) {
    result['@extractedAt'] = metadata.extractedAt;
  }
  if (metadata.method !== undefined) {
    result['@method'] = metadata.method;
  }
  if (metadata.humanVerified !== undefined) {
    result[KEYWORD_HUMAN_VERIFIED] = metadata.humanVerified;
  }
  if (metadata.mediaType !== undefined) {
    result[KEYWORD_MEDIA_TYPE] = metadata.mediaType;
  }
  if (metadata.contentUrl !== undefined) {
    result[KEYWORD_CONTENT_URL] = metadata.contentUrl;
  }
  if (metadata.contentHash !== undefined) {
    result[KEYWORD_CONTENT_HASH] = metadata.contentHash;
  }
  if (metadata.translatedFrom !== undefined) {
    result[KEYWORD_TRANSLATED_FROM] = metadata.translatedFrom;
  }
  if (metadata.translationModel !== undefined) {
    result[KEYWORD_TRANSLATION_MODEL] = metadata.translationModel;
  }
  if (metadata.measurementUncertainty !== undefined) {
    result[KEYWORD_MEASUREMENT_UNCERTAINTY] = metadata.measurementUncertainty;
  }
  if (metadata.unit !== undefined) {
    result[KEYWORD_UNIT] = metadata.unit;
  }
  if (metadata.actedOnBehalfOf !== undefined) {
    result[KEYWORD_ACTED_ON_BEHALF_OF] = metadata.actedOnBehalfOf;
  }
  if (metadata.wasDerivedFrom !== undefined) {
    result[KEYWORD_WAS_DERIVED_FROM] = metadata.wasDerivedFrom;
  }
  if (metadata.wasInvalidatedBy !== undefined) {
    result[KEYWORD_WAS_INVALIDATED_BY] = metadata.wasInvalidatedBy;
  }
  if (metadata.calibrationStatus !== undefined) {
    result[KEYWORD_CALIBRATION_STATUS] = metadata.calibrationStatus;
  }
  if (metadata.calibrationTarget !== undefined) {
    result[KEYWORD_CALIBRATION_TARGET] = metadata.calibrationTarget;
  }
  if (metadata.wasAggregatedBy !== undefined) {
    result[KEYWORD_WAS_AGGREGATED_BY] = metadata.wasAggregatedBy;
  }

  return result;
}

// ── Extraction Helpers ────────────────────────────────────────────

/**
 * Extracts the confidence score from an annotated value or expanded node.
 * Returns undefined if no confidence is present.
 */
export function getConfidence(node: any): number | undefined {
  if (node == null) return undefined;

  // Direct annotation (compact form)
  if (typeof node === 'object' && KEYWORD_CONFIDENCE in node) {
    return node[KEYWORD_CONFIDENCE];
  }

  // Expanded form — look for namespace-qualified key
  const expandedKey = `${JSONLD_EX_NAMESPACE}confidence`;
  if (typeof node === 'object' && expandedKey in node) {
    const val = node[expandedKey];
    if (Array.isArray(val) && val.length > 0) {
      return val[0]['@value'] ?? val[0];
    }
    return val?.['@value'] ?? val;
  }

  return undefined;
}

/**
 * Extracts all provenance metadata from an annotated value.
 */
export function getProvenance(node: any): ProvenanceMetadata {
  const meta: ProvenanceMetadata = {};

  if (node == null || typeof node !== 'object') return meta;

  // Try compact form first, then expanded
  meta.confidence = extractField(node, 'confidence', KEYWORD_CONFIDENCE);
  meta.source = extractField(node, 'source', KEYWORD_SOURCE);
  meta.extractedAt = extractField(node, 'extractedAt', KEYWORD_EXTRACTED_AT);
  meta.method = extractField(node, 'method', KEYWORD_METHOD);
  meta.humanVerified = extractField(node, 'humanVerified', KEYWORD_HUMAN_VERIFIED);
  meta.mediaType = extractField(node, 'mediaType', KEYWORD_MEDIA_TYPE);
  if (!meta.mediaType && node['@type']) {
    meta.mediaType = node['@type'];
  }
  meta.contentUrl = extractField(node, 'contentUrl', KEYWORD_CONTENT_URL);
  meta.contentHash = extractField(node, 'contentHash', KEYWORD_CONTENT_HASH);
  meta.translatedFrom = extractField(node, 'translatedFrom', KEYWORD_TRANSLATED_FROM);
  meta.translationModel = extractField(node, 'translationModel', KEYWORD_TRANSLATION_MODEL);
  meta.measurementUncertainty = extractField(node, 'measurementUncertainty', KEYWORD_MEASUREMENT_UNCERTAINTY);
  meta.unit = extractField(node, 'unit', KEYWORD_UNIT);
  meta.actedOnBehalfOf = extractField(node, 'actedOnBehalfOf', KEYWORD_ACTED_ON_BEHALF_OF);
  meta.wasDerivedFrom = extractField(node, 'wasDerivedFrom', KEYWORD_WAS_DERIVED_FROM);
  meta.wasInvalidatedBy = extractField(node, 'wasInvalidatedBy', KEYWORD_WAS_INVALIDATED_BY);
  meta.calibrationStatus = extractField(node, 'calibrationStatus', KEYWORD_CALIBRATION_STATUS);
  meta.calibrationTarget = extractField(node, 'calibrationTarget', KEYWORD_CALIBRATION_TARGET);
  meta.wasAggregatedBy = extractField(node, 'wasAggregatedBy', KEYWORD_WAS_AGGREGATED_BY);

  // Remove undefined entries
  return Object.fromEntries(
    Object.entries(meta).filter(([_, v]) => v !== undefined)
  ) as ProvenanceMetadata;
}

/**
 * Filters a graph to only include nodes meeting a minimum confidence threshold.
 *
 * @param graph - Array of expanded JSON-LD nodes
 * @param property - Property name to check confidence on
 * @param minConfidence - Minimum confidence threshold (0.0 - 1.0)
 * @returns Filtered array of nodes
 */
export function filterByConfidence(
  graph: any[],
  property: string,
  minConfidence: number
): any[] {
  validateConfidence(minConfidence);

  return graph.filter((node) => {
    const propValue = node[property];
    if (!propValue) return false;

    const values = Array.isArray(propValue) ? propValue : [propValue];
    return values.some((v: any) => {
      const conf = getConfidence(v);
      return conf !== undefined && conf >= minConfidence;
    });
  });
}

/**
 * Aggregates confidence scores across multiple sources using
 * configurable strategies.
 */
export function aggregateConfidence(
  scores: number[],
  strategy: 'mean' | 'max' | 'min' | 'weighted' = 'mean',
  weights?: number[]
): number {
  if (scores.length === 0) return 0;
  scores.forEach(validateConfidence);

  switch (strategy) {
    case 'max':
      return Math.max(...scores);
    case 'min':
      return Math.min(...scores);
    case 'weighted': {
      if (!weights || weights.length !== scores.length) {
        throw new Error('Weights array must match scores array length');
      }
      const totalWeight = weights.reduce((a, b) => a + b, 0);
      const weighted = scores.reduce((sum, s, i) => sum + s * weights[i], 0);
      return weighted / totalWeight;
    }
    case 'mean':
    default:
      return scores.reduce((a, b) => a + b, 0) / scores.length;
  }
}

// ── Internal Helpers ──────────────────────────────────────────────

function validateConfidence(score: number): void {
  if (typeof score !== 'number' || score < 0 || score > 1) {
    throw new RangeError(
      `@confidence must be a number between 0.0 and 1.0, got: ${score}`
    );
  }
}

function extractField(
  node: any,
  compactName: string,
  keyword: string
): any {
  // Compact form (using @-keyword)
  if (keyword in node) {
    return node[keyword];
  }
  // Compact form (using term name)
  if (compactName in node) {
    return node[compactName];
  }
  // Expanded form (namespace-qualified)
  const expandedKey = `${JSONLD_EX_NAMESPACE}${compactName}`;
  if (expandedKey in node) {
    const val = node[expandedKey];
    if (Array.isArray(val) && val.length > 0) {
      return val[0]['@value'] ?? val[0];
    }
    return val?.['@value'] ?? val;
  }
  return undefined;
}

File: packages/python/src/jsonld_ex/ai_ml.py
"""
AI/ML Extensions for JSON-LD

Provides @confidence, @source, @extractedAt, @method, @humanVerified
"""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Any, Optional, Sequence, Literal
import math

JSONLD_EX_NAMESPACE = "http://www.w3.org/ns/jsonld-ex/"


@dataclass
class ProvenanceMetadata:
    """AI/ML provenance metadata attached to a value."""
    confidence: Optional[float] = None
    source: Optional[str] = None
    extracted_at: Optional[str] = None
    method: Optional[str] = None
    human_verified: Optional[bool] = None
    # Multimodal (GAP-MM1)
    media_type: Optional[str] = None
    content_url: Optional[str] = None
    content_hash: Optional[str] = None
    # Translation provenance (GAP-ML2)
    translated_from: Optional[str] = None
    translation_model: Optional[str] = None
    # Measurement uncertainty (GAP-IOT1)
    measurement_uncertainty: Optional[float] = None
    unit: Optional[str] = None


def annotate(
    value: Any,
    confidence: Optional[float] = None,
    source: Optional[str] = None,
    extracted_at: Optional[str] = None,
    method: Optional[str] = None,
    human_verified: Optional[bool] = None,
    # Multimodal (GAP-MM1)
    media_type: Optional[str] = None,
    content_url: Optional[str] = None,
    content_hash: Optional[str] = None,
    # Translation provenance (GAP-ML2)
    translated_from: Optional[str] = None,
    translation_model: Optional[str] = None,
    # Measurement uncertainty (GAP-IOT1)
    measurement_uncertainty: Optional[float] = None,
    unit: Optional[str] = None,
) -> dict[str, Any]:
    """Create an annotated JSON-LD value with provenance metadata."""
    result: dict[str, Any] = {"@value": value}

    if confidence is not None:
        _validate_confidence(confidence)
        result["@confidence"] = confidence
    if source is not None:
        result["@source"] = source
    if extracted_at is not None:
        result["@extractedAt"] = extracted_at
    if method is not None:
        result["@method"] = method
    if human_verified is not None:
        result["@humanVerified"] = human_verified
    # Multimodal
    if media_type is not None:
        result["@mediaType"] = media_type
    if content_url is not None:
        result["@contentUrl"] = content_url
    if content_hash is not None:
        result["@contentHash"] = content_hash
    # Translation
    if translated_from is not None:
        result["@translatedFrom"] = translated_from
    if translation_model is not None:
        result["@translationModel"] = translation_model
    # Measurement
    if measurement_uncertainty is not None:
        result["@measurementUncertainty"] = measurement_uncertainty
    if unit is not None:
        result["@unit"] = unit

    return result


def get_confidence(node: Any) -> Optional[float]:
    """Extract confidence score from a node or annotated value."""
    if node is None or not isinstance(node, dict):
        return None

    # Compact form
    if "@confidence" in node:
        return node["@confidence"]

    # Expanded form
    key = f"{JSONLD_EX_NAMESPACE}confidence"
    if key in node:
        val = node[key]
        if isinstance(val, list) and len(val) > 0:
            item = val[0]
            return item.get("@value", item) if isinstance(item, dict) else item
        if isinstance(val, dict):
            return val.get("@value", val)
        return val

    return None


def get_provenance(node: Any) -> ProvenanceMetadata:
    """Extract all provenance metadata from a node."""
    if node is None or not isinstance(node, dict):
        return ProvenanceMetadata()

    return ProvenanceMetadata(
        confidence=_extract_field(node, "confidence", "@confidence"),
        source=_extract_field(node, "source", "@source"),
        extracted_at=_extract_field(node, "extractedAt", "@extractedAt"),
        method=_extract_field(node, "method", "@method"),
        human_verified=_extract_field(node, "humanVerified", "@humanVerified"),
        # Multimodal
        media_type=_extract_field(node, "mediaType", "@mediaType"),
        content_url=_extract_field(node, "contentUrl", "@contentUrl"),
        content_hash=_extract_field(node, "contentHash", "@contentHash"),
        # Translation
        translated_from=_extract_field(node, "translatedFrom", "@translatedFrom"),
        translation_model=_extract_field(node, "translationModel", "@translationModel"),
        # Measurement
        measurement_uncertainty=_extract_field(node, "measurementUncertainty", "@measurementUncertainty"),
        unit=_extract_field(node, "unit", "@unit"),
    )


def filter_by_confidence(
    graph: Sequence[dict[str, Any]],
    property_name: str,
    min_confidence: float,
) -> list[dict[str, Any]]:
    """Filter graph nodes by minimum confidence on a property."""
    _validate_confidence(min_confidence)
    results = []
    for node in graph:
        prop_value = node.get(property_name)
        if prop_value is None:
            continue
        values = prop_value if isinstance(prop_value, list) else [prop_value]
        if any(
            (c := get_confidence(v)) is not None and c >= min_confidence
            for v in values
        ):
            results.append(node)
    return results


def aggregate_confidence(
    scores: Sequence[float],
    strategy: Literal["mean", "max", "min", "weighted"] = "mean",
    weights: Optional[Sequence[float]] = None,
) -> float:
    """Aggregate multiple confidence scores."""
    if len(scores) == 0:
        return 0.0
    for s in scores:
        _validate_confidence(s)

    if strategy == "max":
        return max(scores)
    elif strategy == "min":
        return min(scores)
    elif strategy == "weighted":
        if weights is None or len(weights) != len(scores):
            raise ValueError("Weights must match scores length")
        for w in weights:
            if not isinstance(w, (int, float)) or isinstance(w, bool):
                raise TypeError(f"Weight must be a number, got: {type(w).__name__}")
            if w < 0:
                raise ValueError(f"Weight must be non-negative, got: {w}")
        total_weight = sum(weights)
        if total_weight == 0:
            raise ValueError("Total weight must be greater than zero")
        return sum(s * w for s, w in zip(scores, weights)) / total_weight
    else:  # mean
        return sum(scores) / len(scores)


# ── Internal ───────────────────────────────────────────────────────

def _validate_confidence(score: float) -> None:
    if not isinstance(score, (int, float)) or isinstance(score, bool):
        raise TypeError(f"@confidence must be a number, got: {type(score).__name__}")
    if math.isnan(score) or math.isinf(score):
        raise ValueError(f"@confidence must be finite, got: {score}")
    if score < 0 or score > 1:
        raise ValueError(f"@confidence must be between 0.0 and 1.0, got: {score}")


def _extract_field(node: dict, compact_name: str, keyword: str) -> Any:
    if keyword in node:
        return node[keyword]
    if compact_name in node:
        return node[compact_name]
    expanded_key = f"{JSONLD_EX_NAMESPACE}{compact_name}"
    if expanded_key in node:
        val = node[expanded_key]
        if isinstance(val, list) and len(val) > 0:
            item = val[0]
            return item.get("@value", item) if isinstance(item, dict) else item
        if isinstance(val, dict):
            return val.get("@value", val)
        return val
    return None
